source/core/DualTypeVisitor.cpp

```diff
@@ -0,0 +1,323 @@
+                case kEncoding_Boolean:
+                    success = operation->BooleanCompatible(typeRefX, typeRefY);
+                    break;
+                case kEncoding_UInt:
+                    success = operation->UIntCompatible(typeRefX, typeRefY);
+                    break;
+                case kEncoding_S2CInt:
+                    success = operation->S2CIntCompatible(typeRefX, typeRefY);
+                    break;
+                case kEncoding_IEEE754Binary:
+                    success = operation->IEEE754BinaryCompatible(typeRefX, typeRefY);
+                    break;
+                case kEncoding_Cluster:
+                    success = ClusterCompatible(typeRefX, pDataX, typeRefY, pDataY, operation);
+                    break;
+                case kEncoding_Enum:
+                    success = EnumCompatible(typeRefX, typeRefY, operation);
+                    break;
+                case kEncoding_Array: {
+                    if (typeRefX->Rank() == 1 && typeRefX->GetSubElement(0)->BitEncoding() == kEncoding_Unicode) {
+                        success = StringCompatible(typeRefX, typeRefY);
+                    } else {
+                        success = ArrayCompatible(typeRefX, typeRefY, operation);
+                    }
+                    break;
+                }
+                default:
+                    success = false;
+            }
+        }
+        return success;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeVisitor::VariantCompatible(TypeRef typeRefX, void* pDataX, TypeRef typeRefY, void* pDataY, DualTypeOperation* operation)
+    {
+        VariantTypeRef variantTypeX = reinterpret_cast<VariantTypeRef>(typeRefX);
+        VariantTypeRef variantTypeY = reinterpret_cast<VariantTypeRef>(typeRefY);
+
+        TypeRef variantUnderlyingTypeX = variantTypeX->_underlyingTypeRef;
+        TypeRef variantUnderlyingTypeY = variantTypeY->_underlyingTypeRef;
+        bool success = false;
+        if (!variantUnderlyingTypeX && !variantUnderlyingTypeY) {
+            success = true;
+        } else if (!variantUnderlyingTypeX || !variantUnderlyingTypeY) {
+            return false;
+        } else {
+            success = TypesAreCompatible(variantUnderlyingTypeX,
+                variantUnderlyingTypeX->Begin(kPARead),
+                variantUnderlyingTypeY,
+                variantUnderlyingTypeY->Begin(kPARead),
+                operation);
+        }
+        if (success) {
+            // compare attributes
+            if (!variantTypeX->_attributeMap && !variantTypeY->_attributeMap) {
+                return true;
+            } else if (variantTypeX->_attributeMap && variantTypeY->_attributeMap) {
+                if (variantTypeX->_attributeMap->size() != variantTypeY->_attributeMap->size()) {
+                    return false;
+                }
+                for (const auto attributePairInX : *variantTypeX->_attributeMap) {
```

Comment:
Since traversal order of std::map is ordered, iterating both attribute maps together in one loop will be more efficient. However, it is ok not to do this in the PR. Not a bad idea to leave a comment though.